contracts/Buffer.sol

```diff
@@ -0,0 +1,169 @@
+  * @return The final position of the cursor (will equal `_offset` if `_relative` is `false`).
+  */
+  function seek(Buffer memory _buffer, uint64 _offset, bool _relative) internal pure returns (uint64) {
+    uint64 newCursor = _offset;
+    // Deal with relative offsets
+    if (_relative == true) {
+      newCursor += _buffer.cursor;
+    }
+    // Make sure not to read out of the bounds of the original bytes
+    require(newCursor < _buffer.data.length, "Not enough bytes in buffer when seeking");
+    _buffer.cursor = newCursor;
+    return _buffer.cursor;
+  }
+
+  /**
+  * @notice Move the inner cursor a number of bytes forward.
+  * @dev This is a simple wrapper around the relative offset case of `seek()`.
+  * @param _buffer An instance of `BufferLib.Buffer`.
+  * @param _relativeOffset How many bytes to move the cursor forward.
+  * @return The final position of the cursor.
+  */
+  function seek(Buffer memory _buffer, uint64 _relativeOffset) internal pure returns (uint64) {
+    return seek(_buffer, _relativeOffset, true);
+  }
+
+  /**
+  * @notice Move the inner cursor back to the first byte in the buffer.
+  * @param _buffer An instance of `BufferLib.Buffer`.
+  */
+  function rewind(Buffer memory _buffer) internal pure {
+    _buffer.cursor = 0;
+  }
+
+  /**
+  * @notice Read and consume the next byte from the buffer as an `uint8`.
+  * @param _buffer An instance of `BufferLib.Buffer`.
+  * @return The `uint8` value of the next byte in the buffer counting from the cursor position.
+  */
+  function readUint8(Buffer memory _buffer) internal pure returns (uint8) {
+    return uint8(read(_buffer, 1)[0]);
+  }
+
+  /**
+  * @notice Read and consume the next 2 bytes from the buffer as an `uint16`.
+  * @param _buffer An instance of `BufferLib.Buffer`.
+  * @return The `uint16` value of the next 2 bytes in the buffer counting from the cursor position.
+  */
+  function readUint16(Buffer memory _buffer) internal pure returns (uint16) {
+    bytes memory bytesValue = read(_buffer, 2);
+    return (uint16(uint8(bytesValue[0])) << 8) |
+      uint8(bytesValue[1]);
+  }
+
+  /**
+  * @notice Read and consume the next 4 bytes from the buffer as an `uint32`.
+  * @param _buffer An instance of `BufferLib.Buffer`.
+  * @return The `uint32` value of the next 4 bytes in the buffer counting from the cursor position.
+  */
+  function readUint32(Buffer memory _buffer) internal pure returns (uint32) {
+    bytes memory bytesValue = read(_buffer, 4);
+    return (uint32(uint8(bytesValue[0])) << 24) |
+      (uint32(uint8(bytesValue[1])) << 16) |
+      (uint16(uint8(bytesValue[2])) << 8) |
+      uint8(bytesValue[3]);
+  }
+
+  /**
+  * @notice Read and consume the next 8 bytes from the buffer as an `uint64`.
+  * @param _buffer An instance of `BufferLib.Buffer`.
+  * @return The `uint64` value of the next 4 bytes in the buffer counting from the cursor position.
+  */
+  function readUint64(Buffer memory _buffer) internal pure returns (uint64) {
```

Comment:
As we saw together today, these functions to read from bytes are much more efficient in assembly. :blush: